libs/point_to_point_frr_helper.h

```diff
@@ -0,0 +1,387 @@
+
+    NetDeviceContainer Install(std::string aName, Ptr<Node> b);
+
+    NetDeviceContainer Install(std::string aNode, std::string bNode);
+
+  private:
+    void EnablePcapInternal(std::string prefix, Ptr<NetDevice> nd,
+                            bool promiscuous, bool explicitFilename) override;
+
+    void EnableAsciiInternal(Ptr<OutputStreamWrapper> stream,
+                             std::string prefix, Ptr<NetDevice> nd,
+                             bool explicitFilename) override;
+
+    ObjectFactory m_queueFactory;
+    ObjectFactory m_channelFactory;
+    ObjectFactory m_deviceFactory;
+    bool m_enableFlowControl;
+};
+
+/***************************************************************
+ *  Implementation of the templates declared above.
+ ***************************************************************/
+
+template <typename FRR_POLICY>
+template <typename... Ts>
+void PointToPointFRRHelper<FRR_POLICY>::SetQueue(std::string type, Ts&&... args)
+{
+    QueueBase::AppendItemTypeIfNotPresent(type, "Packet");
+
+    m_queueFactory.SetTypeId(type);
+    m_queueFactory.Set(std::forward<Ts>(args)...);
+}
+
+template <typename FRR_POLICY>
+PointToPointFRRHelper<FRR_POLICY>::PointToPointFRRHelper()
+{
+    m_queueFactory.SetTypeId("ns3::DropTailQueue<Packet>");
+    m_deviceFactory.SetTypeId(
+        ns3::PointToPointFRRNetDevice<FRR_POLICY>::getNetDeviceString());
+    m_channelFactory.SetTypeId(
+        ns3::PointToPointFRRChannel<FRR_POLICY>::getChannelString());
+    m_enableFlowControl = true;
+}
+
+template <typename FRR_POLICY>
+void PointToPointFRRHelper<FRR_POLICY>::SetDeviceAttribute(
+    std::string n1, const AttributeValue& v1)
+{
+    m_deviceFactory.Set(n1, v1);
+}
+
+template <typename FRR_POLICY>
+void PointToPointFRRHelper<FRR_POLICY>::SetChannelAttribute(
+    std::string n1, const AttributeValue& v1)
+{
+    m_channelFactory.Set(n1, v1);
+}
+
+template <typename FRR_POLICY>
+void PointToPointFRRHelper<FRR_POLICY>::DisableFlowControl()
+{
+    m_enableFlowControl = false;
+}
+
+template <typename FRR_POLICY>
+void PointToPointFRRHelper<FRR_POLICY>::EnablePcapInternal(
+    std::string prefix, Ptr<NetDevice> nd, bool promiscuous,
+    bool explicitFilename)
+{
+    //
+    // All of the Pcap enable functions vector through here including the ones
+    // that are wandering through all of devices on perhaps all of the nodes in
+    // the system.  We can only deal with devices of type
+    // PointToPointFRRNetDevice.
+    //
+    Ptr<PointToPointFRRNetDevice<FRR_POLICY>> device =
+        nd->GetObject<PointToPointFRRNetDevice<FRR_POLICY>>();
+    if (!device) {
+        NS_LOG_INFO("Device " << device
+                              << " not of type ns3::PointToPointFRRNetDevice");
+        return;
+    }
+
+    PcapHelper pcapHelper;
+
+    std::string filename;
+    if (explicitFilename) {
+        filename = prefix;
+    } else {
+        filename = pcapHelper.GetFilenameFromDevice(prefix, device);
```

Comment:
Does this mean we can mess with the file names a bit? Might help to clear up the different suffixes to know which gateway we are looking at.